File: algorithms/algorithms-python/leetcode/Question_0017_Letter_Combinations_of_a_Phone_Number.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        res = []
        pre = []
        dic = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        self.dfs(digits, dic, 0, pre, res)
        return res

    def dfs(self, digits, dic, level, pre, res):
        if level == len(digits):
            res.append("".join(pre))
        else:
            s = dic[int(digits[level]) - 2]
            for c in s:
                pre.append(c)
                self.dfs(digits, dic, level + 1, pre, res)
                pre.pop()

File: algorithms/algorithms-cplusplus/leetcode/Question_0017_Letter_Combinations_of_a_Phone_Number.cc
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<string> letterCombinations(string digits) {
    vector<string> res;
    if (digits.empty()) {
      return res;
    }
    vector<string> dict = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    string s = "";
    dfs(digits, dict, 0, s, res);
    return res;
  }

  void dfs(string& digits, vector<string>& dict, int i, string& s, vector<string>& res) {
    if (i >= digits.size()) {
      res.push_back(s);
      return;
    }
    for (char c : dict[digits[i] - '2']) {
      s.push_back(c);
      dfs(digits, dict, i + 1, s, res);
      s.pop_back();
    }
  }
};
